### ISSUE TITLE
Blazor: bind-value-oninput or bind does not work with contenteditable divs

### ISSUE THREAD

**Issue by PirasannaRavi** [dotnet/aspnetcore#9974]
### Bug Description
In the below code, the expectation is as and when I edit the div content I need to update the model so that I can then save it back to Database.

### To Reproduce
1. Using the latest preview version of ASP.NET Core 3
2. Run the below code 
`<div contenteditable="true" bind-value-oninput="@contentEditableText">`
                                               `You can Edit this Text`
                                            `</div>`
                                           ` @contentEditableText`
`@functions {
    string contentEditableText = "You can Edit this Text";
}`
4. As I Type something in the editable div, I need to see the same reflected in the body, but it does not happen that way. The OnChange event does not fire automatically.

### Expected behavior
OnChange Event needs to fire and changes in editable div should be bound to the model.

### Screenshots
<img width="207" alt="Screen Shot 2019-05-04 at 2 01 06 PM" src="https://user-images.githubusercontent.com/49619395/57183069-90bc4f80-6e75-11e9-8ad7-5130480f2faa.png">
<img width="380" alt="Screen Shot 2019-05-04 at 2 02 50 PM" src="https://user-images.githubusercontent.com/49619395/57183070-90bc4f80-6e75-11e9-86e8-ae4dd11038ff.png">

**Comment by SteveSandersonMS** [comment]
We don't currently support `bind` with `contenteditable`. It might be something we add in the future, but it's not directly planned.

In the meantime, you can use `contenteditable` in conjunction with JS interop to receive notification when edits occur and update your .NET models.

**Comment by SteveSandersonMS** [comment]
Reopening in case this is something we choose to address in the future.

**Comment by Andrzej-W** [comment]
@SteveSandersonMS please look at #10087. It looks that Blazor renders incorrect HTML when we try to use `contenteditable` with simple JS interop.

**Comment by lfdgaliza** [comment]
I will give another suggestion here just in case there is someone else with the same problem that I had.
In my case, I needed to create inputs dynamically, so I could not just bind the contenteditable span using @ref to only one ElementReference.

Then I made this:

```csharp
Dictionary<string, ElementReference> References; // Some null check boilerplate omitted

if(!References.ContainsKey(word))
    References.Add(word, default(ElementReference));

<span @ref="References[word]" class="word" contenteditable="true" @onfocusout="(e) => Word_FocusOut(word)" ></span>
```
Then in my Word_FocusOut, I can have access to the correct ElementReference by accessing References[word]:

```csharp
async void Word_FocusOut(string word)
{
    var text = await jsRuntime.InvokeAsync<string>("JsLib.getInnerHtml", References[word]);
    Console.WriteLine(text);
}
```

Please let me know if there is a better way to address this.

Thanks!

**Comment by neroqfi** [comment]
Hello,

if you're planning to implement contenteditable div binding, I addressed an issue few weeks ago related to this (FYI):

https://developercommunity.visualstudio.com/content/problem/1175416/blazor-webassembly-runtimeerror-memory-access-out.html

Maybe the contenteditable div binding could resolve my issue also? There comes restrictions when I try to fetch innerHTML from the C# by using JS interop.

**Comment by kkentt** [comment]
> Reopening in case this is something we choose to address in the future.

That would be great if this feature is supported in future releases.

**Comment by MaxFlawless** [comment]
I agree definitely a must-have feature textarea is getting killed! contenteditable is so much more user-friendly

**Comment by kefyru** [comment]
add js script:
```js
...
document.addEventListener("input", onInput);

function onInput(e) {
  let target = e.target;
  
  if (!target.value && !target.__contenteditable) target.__contenteditable = true;
  
  if (target.__contenteditable) target.value = target.innerText;
}
...
```

**Comment by EislerDavid** [comment]
> kefyru

Your code breaks textareas and inputs. (Try to write somethiong in the textarea, then delete it all. After that is textarea broken.)

Here is my fix:
 ```
document.addEventListener("input", onInput);
function onInput(e) {
  let target = e.target;
  if (target.localName == "div") {
    if (!target.value && !target.__contenteditable) target.__contenteditable = true;
    if (target.__contenteditable) target.value = target.innerText;
  }
}
 ```

**Comment by TomoPeeka** [comment]
Any news on this matter?

**Comment by ersk** [comment]
I use contenteditable all the time. Definitely a feature I would like.

**Comment by BieleckiLtd** [comment]
> > kefyru
how does one use your js?

**Comment by ShaunCurtis** [comment]
The examples I see above are "Happy Path" examples.  I started to look at question on SO about this: again a nice simple happy path example given.

I can see using `contenteditable` in an `InputBase` or similar controlled environment.  But as an attribute?  

Here's some examples going from happy path to very edge conditions.     There's the `RenderFragment` problem - code between a `<div>` and `</div>` in a Razor file is built as a `RenderFragment` not text.  And how would you start trying to decipher the content in the Edge example????

```csharp
@page "/"

<PageTitle>Index</PageTitle>

<h3 class="text-secondary border-bottom border-secondary">Happy Path</h3>
<div contenteditable="true" class="m-3 p-3 bg-light border border-primary">
    Welcome to your new app.
</div>

<h3 class="text-secondary border-bottom border-secondary">More Edge</h3>
<div contenteditable="true" class="m-3 p-3 bg-light border border-primary">
    <h1>Hello, world!</h1>

    Welcome to your new app.
</div>

<h3 class="text-secondary border-bottom border-secondary">Very Edge!</h3>
<div contenteditable="true" class="m-3 p-3 bg-light border border-primary">
    <EditForm Model=this.model>
        Value: <InputText @bind-Value=model.Value />
        <button type="submit">Submit</button>
    </EditForm>
</div>

@code {
    private DataModel model = new DataModel();

    public class DataModel
    {
        public string Value { get; set; } = string.Empty;
    }
}
```

**Comment by drocha87** [comment]
I would like to share a demo repository that I have where I used a `contenteditable` to implement a mention text area. This is not the `very edge` use case as suggested by @ShaunCurtis but I think it's a little bit more complex than the "happy path". 

You'll realized that I used a lot of JS to handle the editor, this is because my project was a Blazor Server Side so I need to manage the editor on client side for obvious reasons. I'm not sure if I'm doing things the right way, but it ended up working properly. 

[Blazor Server Mentions Repository](https://github.com/drocha87/blazorServerMentions)

You may be interested in the [component](https://github.com/drocha87/blazorServerMentions/tree/main/Components) and in the [JS Interop](https://github.com/drocha87/blazorServerMentions/blob/main/TScripts/editor.ts#L40).

**Comment by Andrzej-W** [comment]
Based on the number of thumbs up I think a lot of people (including me) are interested in this feature. What should we do to prioritize this work?

**Comment by vukasinpetrovic** [comment]
I just came across this issue while searching for a solution. I was hoping so much that it would be included in .net 8 release, while I was reading the comments here. 
I tried the proposed solution with JS function, but for some reason I did not manage to make it work and trigger second way of two way binding. I'm trying to find solution without using some component library to have "inline edit" feature for some label/text element. 

For example here, if I change the input field value, it reflects and changes the displayed section name in both input field and editable div. But if I try to change value in div (via contenteditable), it does not trigger the value change and thus it does not change SectionName variable value. I set the breakpoint in that JS, that code is reached, but no changes are made to the model variable. Am I doing something wrong?
```razor
<div contenteditable="true" placeholder="Section Name" bind-value="@SectionName"></div>
<InputText @bind-Value="SectionName" placeholder="Section Name"/>
```

```js
<script>
        document.addEventListener("input", onInput);
        function onInput(e) {
            let target = e.target;
            if (target.localName == "div") {
                if (!target.value && !target.__contenteditable) target.__contenteditable = true;
                if (target.__contenteditable) target.value = target.innerText;
            }
        }
    </script>
```

**Comment by vukasinpetrovic** [comment]
I tried now with this combination, where it reflects the changes from div into input, but it does not work the other way around

```razor
<div contenteditable="true" placeholder="Section Name" @bind-value="SectionName" @bind-value:event="oninput"></div>
<InputText @bind-Value="SectionName" placeholder="Section Name"/>
```

**Comment by vukasinpetrovic** [comment]
Ok, another update, in order to display the value inside the div, I added <div>@SectionName</div> so the variable is it's inner content. Now two way binding and all the changes are reflected on both the div and the input. 
BUT, when I change text value inside div, after every input (key press) it resets the cursor to the beginning of the content. This is probably because Blazor is re-rendering variable which is set as content of the div. I'm not sure how to solve this. Any help is appreciated. 

P.S. I'm basically trying to implement in-place editor this way. Yes there are components for it, but with contenteditable is the most user friendly for my need, as it prevents the width and height before, while and after editing.

**Comment by EngieDev** [comment]
@vukasinpetrovic Maybe you could use the `@key` attribute to prevent re-rendering?

**Comment by luetm** [comment]
Would really be great. It would let us get around annoying limitations and behaviors of `<input>` in html/css (e.g. that you can't size them to content, they flex/grow weirdly, ...). An editable `div` is an elegant workaround for this. Especially with Blazor Server, the workaround with JS is really annoying unfortuately.

**Comment by FLAMESpl** [comment]
Support for contenteditable divs would be a great opportunity for implementing blazor-native WYSIWYG editors.
